crates/sdk/src/websocket.rs

```diff
@@ -313,15 +355,42 @@ impl WsConnection {
                     }
 
                     Ok(Some(WebSocketMessage::Ping(payload))) => {
+                        log::trace!("received ping");
+                        idle = false;
                         record_metrics(payload.len());
-                    }
+                    },
+
+                    Ok(Some(WebSocketMessage::Pong(payload))) => {
+                        log::trace!("received pong");
+                        idle = false;
+                        want_pong = false;
+                        record_metrics(payload.len());
+                    },
 
                     Ok(Some(other)) => {
                         log::warn!("Unexpected WebSocket message {:?}", other);
+                        idle = false;
                         record_metrics(other.len());
                     },
                 },
 
+                _ = idle_timeout_interval.tick() => {
+                    if mem::replace(&mut idle, true) {
+                        if mem::take(&mut want_pong) {
```

Comment:
Why are we `mem::take`ing here? It looks to me like this is equivalent to just reading `want_pong` in all cases.

- `want_pong == true`: `take` and replace with `false`. Break out of the loop and never read from `want_pong` again.
- `want_pong == false`: `take` is a no-op. Set `want_pong` to `true` later in this block. Never read the `false` value installed by `mem::take`.

It seems like there are two possible simplifications to this code:
- `mem::replace(&mut want_pong, true)` here, and remove the `want_pong = true` assignment on line 390.
- Just branch on `want_pong` with `if want_pong {` here, and leave the assignment later on.